### DISCUSSION TITLE
Implicit major updates

### DISCUSSION THREAD

**Discussion by mbrodala** [renovatebot/renovate#13993] [Request Help] [answered]
Can Renovate prevent implicit major updates?

Example:

- Composer/Symfony project
- Requires `"symfony/security-bundle": "^5.2"` (currently at version `5.4.2`)
- Installs `symfony/security-core` (currently at version `5.4.2`) but does not require it explicitly, since this is already required implicitly by `symfony/security-bundle`

Among others, Renovate will suggest an update of `symfony/security-bundle` from 5.4.2 to 5.4.3 which is perfectly fine.

However, since `symfony/security-bundle` allows `"symfony/security-core": "^5.4|^6.0"`, the latter is implicitly also updated from `5.4.2` to `6.0.3`. This brings breaking changes obviously. (In this case a [change to the `PasswordUpgraderInterface`](https://github.com/symfony/symfony/commit/30e2c008b4cfd6eefdb278fb11e67945730db057))

Now the question once more: can Renovate prevent update suggestions like these?

Technically all [Symfony packages are grouped](https://docs.renovatebot.com/presets-group/#groupsymfony) by default but apparently there is nothing in place which prevents mixed major versions here.

Normally it shouldn't matter, the Symfony packages explicitly allow such a mix after all. But IMO there should be an option at least to ensure patch/minor updates of explicit dependencies never pull major updates of implicit dependencies.

**Comment by rarkins** [comment, marked as answer]
If the question is "Can or should Renovate prevent a direct dependency from bumping a major version of one of its transitive dependencies?" then the answer is definitely not, we should never block that - dependencies are responsible for absorbing breaking changes from their own dependencies.

But I'm not sure what you mean by "Installs symfony/security-core (currently at version 5.4.2) but does not require it explicitly". Do you mean that you use it directly but decided not to put it in your `composer.json` because you knew it's already a transitive dependency?

**Reply by mbrodala** [reply]
Thanks for the reply. Yes, the latter is exactly our current state.

We try to have only minimal dependencies and thus technically rely on implicit package versions. And in fact I plan to fix the issue at hand by an explicit dependency on `symfony/security-core` version 5.x.

But I'm not sure if I open Pandora's box this way and in the end will need to require dozens of packages already installed but only implicitly.

**Reply by rarkins** [reply]
In JS it's considered bad practice to require a package which isn't a direct dependency, and I think it should be the same for other languages too. You should not only be worried about a major update of a depended-upon indirect dependency but also the *removal* of one too. Even though "core" admittedly sounds like one they'll keep, one day they might rewrite their core, call it something like "magnolia", and depend upon that instead.

To answer your question, I don't think that we'd likely support complicated attempts to prevent "implicit" major updates when it could be solved through other means which I think most would agree are best practice.

**Reply by mbrodala** [reply]
Alright, thanks for the clarification. In this case I'll try to gather all the other implicit dependencies which should actually be explicit ones.
